build-std: explicit dependencies

davidtwco opened the pull request:
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rfcs/pull/3875)*

Allow users to add explicit dependencies on standard library crates in the `Cargo.toml`. This enables Cargo to determine which standard library crates are required by the crate graph without `build-std-crates` being set and for different crates to require different standard library crates.

**This RFC is is part of the [build-std project goal](https://rust-lang.github.io/rust-project-goals/2025h2/build-std.html) and a series of build-std RFCs:**

1. build-std context (rust-lang/rfcs#3873)
2. `build-std="always"` (rust-lang/rfcs#3874)
3. Explicit standard library dependencies (this RFC)
4. `build-std="compatible"` (RFC not opened yet)
5. `build-std="match-profile"` (RFC not opened yet)

There is also [a Zulip channel where you can ask questions about any of the build-std RFCs](https://rust-lang.zulipchat.com/#narrow/channel/516120-project-goals.2F2025h1.2Fbuild-std). This series of RFCs was drafted over many months with the help of stakeholders from many Rust project teams, we thank them for their help!

There are some details that have been noted as being worth mentioning in any eventual tracking issues:

- https://github.com/rust-lang/rfcs/pull/3875#discussion_r2482474212

[FCP](https://github.com/rust-lang/rfcs/pull/3875#issuecomment-3715845871)

[Rendered](https://github.com/rust-lang/rfcs/blob/master/text/3875-build-std-explicit-dependencies.md)

davidtwco committed:
build-std: explicit dependencies

davidtwco committed:
rename to rfc 3875

clarfonthey commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1111 @@
+library (e.g. an optional `std` and non-optional `core` or `alloc`, or an
+optional `alloc` and non-optional `core`). `core` cannot be optional. For
+example, the following example will error as it could result in a build without
+`core` (if the `std` feature were disabled):
```
While it is an unstable feature, I think it's worth commenting on how this interacts with the `no_core` feature: will crates using this still have to specify the core dependency, or will it be allowed to be omitted in those cases? Additionally, would this require an additional cargo feature or would just the feature on the crate be enough?

davidtwco replied:
Added some wording around this in 97de9cd5

clarfonthey commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1111 @@
+edition = "2024"
+
+[dependencies]
+std = { builtin = true, optional = true }
```
How does this interact with the `#![no_std]` attribute? Is this attribute no longer required if the `std` dependency is not enabled, or do crates still need to provide it?

epage commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1111 @@
+dependencies to the metadata emitted by `cargo metadata` (when those crates are
+explicit dependencies). None of the standard library's dependencies will be
+included. `source` would be set to `builtin` and the remaining fields would be
+set like any other dependency.
```
Were we doing this only for explicitly builtins and not implicit?

epage replied:
... somehow I missed the parenthetical where this was clarified.

epage commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1111 @@
+> When producing a registry index entry for a package Cargo will not serialise
+> any `builtin` dependencies it inferred. This allows the set of inferred
+> packages to change in the future if needed. Similarly, the published
+> `Cargo.toml` will not explicitly declare any inferred dependencies.
```
This is missing a very important motivation: if we put implicit builtins into `Cargo.toml`, then publishing with a new Cargo would raise your MSRV

davidtwco replied:
Added this motivation in 87c3a95

epage commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1111 @@
+[`cargo add`][cargo-add]'s heuristics will include adding `std`, `alloc` or
+`core` as builtin dependencies if these crate names are provided. `cargo add`
+will additionally have a `--builtin` flag to allow for adding crates with a
+`builtin` source explicitly:
```
Should `cargo add` automatically add `public = true` for builtins?

davidtwco replied:
I've added a note that `public= true` would be added after public/private deps are stabilised in c214948

epage commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1111 @@
+There are crates building on stable which re-export from the standard library.
+If the standard library dependencies were not public then these crates would
+start to trigger the `exported_private_dependencies` lint when upgrading to a
+version of Cargo with a standard library dependency.
```
Is this referring to implicit builtins?  I had read it as any builtin and was confused

epage commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1111 @@
+
+↩ [*Public and private dependencies*][public-and-private-dependencies]
+
+## Why follow the default privacy of explicit standard library dependencies?
```
is this question orphaned?

epage commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1111 @@
+[unresolved-dep-syntax]: #what-syntax-is-used-to-identify-dependencies-on-the-standard-library-in-cargotoml
+
+What syntax should be used for the explicit standard library dependencies?
+`builtin = true`? `sysroot = true`?
```
If we mention `sysroot` can we mention the problem that it has no end-user meaning?

davidtwco committed:
move links to relevant sections

davidtwco committed:
move no-std possibility from previous rfc

davidtwco committed:
reference object form of build-std from prev rfc

davidtwco committed:
mention MSRV-implications of implicit builtins in registry index entries

davidtwco commented on text/3875-build-std-explicit-dependencies.md:
````diff
@@ -0,0 +1,1111 @@
+
+[dependencies]
+std = { builtin = true, public = true }
+```
````
To clarify, a lint in the compiler if we end up passing `--extern std=...` to rustc but with a `#![no_std]` crate? That makes sense to me but the part of the document you've commented on is about public/private dependencies which seems unrelated, so wanted to double check.

clarfonthey replied:
Basically, if:

1. `Cargo.toml` doesn't have any std/alloc/core dependencies
2. The crate is `no_std`

It may require explicit support from Cargo to do this.

clarfonthey replied:
I think that's reasonable; I'll resolve this for now. I definitely think it's okay to punt this off until later.

davidtwco commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1111 @@
+
+[dependencies]
+std = { builtin = true, optional = true }
+# error: must have a non-optional dependency on core
```
I'm not convinced this is that annoying a requirement. Users will only hit this after adding an explicit optional dependency on the `alloc` or `std` and immediately notice the error and need to fix it. I don't see the value in having a lint that would allow you to delay fixing this but would force you to fix it eventually when you want to publish the crate.

clarfonthey replied:
I guess that my main opinion here is that it's something that requires immediate attention to be fixed when it doesn't really affect the ability to build the code. If someone runs `cargo add std --optional`, this suddenly prevents `rust-analyzer` from working at all until it's fixed, and while the error should definitely be fixed before publishing the package, I fail to see how this should completely deadlock cargo's ability to build the crate at all until it's fixed.

davidtwco replied:
> `cargo add` could just automatically add `core` if you're adding `std` or `alloc`.

Added this in 2b189a4c

davidtwco commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1111 @@
+[`cargo add`][cargo-add]'s heuristics will include adding `std`, `alloc` or
+`core` as builtin dependencies if these crate names are provided. `cargo add`
+will additionally have a `--builtin` flag to allow for adding crates with a
+`builtin` source explicitly:
```
I'd expect we'd wait for the natural parse error but I don't feel strongly, if Cargo has different norms around how to deal with cases like this then we can follow those norms.

davidtwco commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1111 @@
+dependencies to the metadata emitted by `cargo metadata` (when those crates are
+explicit dependencies). None of the standard library's dependencies will be
+included. `source` would be set to `builtin` and the remaining fields would be
+set like any other dependency.
```
I think I excluded dependencies of the standard library only as we've tried to hide those as much as possible throughout, but `cargo metadata` is probably a special-case here because it is used by tools like rust-analyzer. I don't know enough about how `cargo metadata` is used by rust-analyzer to know what the right answer here is - do you need to construct a full crate graph including the dependencies of std?

epage replied:
As I said, please leave this as an Unresolved Question

davidtwco committed:
add public with cargo add after stabilisation

davidtwco committed:
clarify implicit implicit not explicit

davidtwco committed:
remove orphaned question

davidtwco committed:
mention sysroot isn't great

davidtwco committed:
unblocks cargo#8798

davidtwco commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1111 @@
+edition = "2024"
+
+[dependencies]
+std = { builtin = true }
```
I think they would continue to use `path`, and then we can simplify how their dependencies refer to them as in [*Remove `rustc_dep_of_std`*](https://github.com/davidtwco/rfcs/blob/build-std-part-three-explicit-dependencies/text/3875-build-std-explicit-dependencies.md#remove-rustc_dep_of_std).

davidtwco committed:
include std deps in cargo metadata

davidtwco commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1111 @@
+context RFC.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
```
Mentioned this in 8996003.

davidtwco committed:
add std deps in cargo metadata to unresolved qs

davidtwco committed:
`no_std` w/ std deps lint in future possiblities

clarfonthey commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1167 @@
+> ├── rand v0.7.3
+> │   ├── getrandom v0.1.14
+> │   │   ├── cfg-if v0.1.10
+> │   │   │   └── core v0.0.0
```
I get why these don't have a usable version number, but IMHO, this should either display as `core (built-in)` or use the current Rust toolchain version instead. I can see the latter being considered very useful for the theoretical long-term case where crates could potentially be compiled with different libstd versions and then linked together, e.g. via dynamic linking.

davidtwco replied:
Removed the version number in fc16fa8f

Ericson2314 commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1167 @@
+
+How should `builtin` deps be represented in lockfiles? Is `builtin = true`
+appropriate? Could the `source` field be reused with the string "builtin" or
+should it stay only as a URL+scheme?
```
I would say the current approach is very good!

A lockfile with such packages should be thought of as a "[Mixin](https://en.wikipedia.org/wiki/Mixin) lockfile" or "[open](https://en.wikipedia.org/wiki/Open_formula) lockfile". When a tool combines such a lockfile with the lockfile for the standard library (according to the perma-unstable patching). Then the "holes" or "placeholders" that are these built-in packages can be filled in, and one is back to having a regular, "closed lockfile" (in memory / in principle, no need to write to disk, necessarily) that doesn't have any such holes.

The ability to combine the lockfiles and then cancel out the new thing bodes well for implementation / conceptual complexity. I means that the new form of lockfile package can be eliminated fairly quickly in Cargo's workflow, at least in principle, and should have minimal effect on the rest of the system.

ehuss committed:
Increase heading levels

As part of https://github.com/rust-lang/rfcs/pull/3883, we have switched
the template to stop using level-1 headings which isn't the way mdbook
is intended to be used.

ehuss commented:
As a heads up, we are planning to start an FCP proposal on this RFC shortly after https://github.com/rust-lang/rfcs/pull/3873 (build-std context) has been approved.

kpreid commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1167 @@
+> Explicit dependencies are passed to rustc without the `noprelude` modifier
+> ([?][rationale-explicit-noprelude]). When adding an explicit dependency, users
+> may need to adjust their code (removing extraneous `extern crate` statements
+> or root-relative paths, like `::std`).
```
That makes sense. I think the text should be revised to make that explicit, because I think that it is not obvious what the text means to people like me, who came to Rust after the release of the 2018 edition.

davidtwco committed:
update links to prev rfcs

davidtwco committed:
built-in instead of version number in cargo tree

davidtwco committed:
adding non-optional core alongside optional std

davidtwco committed:
clarify noprelude meaning

davidtwco committed:
clarify implicit deps

ehuss commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1189 @@
+>
+> When adding an explicit dependency, users > may need to adjust their code
+> (removing extraneous `extern crate` statements > or root-relative paths,
+> like `::std` - this will likely only be the case on the 2015 edition).
```
Looks like there was a word-wrapping issue.


```suggestion
> When adding an explicit dependency, users may need to adjust their code
> (removing extraneous `extern crate` statements or root-relative paths,
> like `::std` - this will likely only be the case on the 2015 edition).
```

epage commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1189 @@
+### Public and private dependencies
+[public-and-private-dependencies]: #public-and-private-dependencies
+
+Implicit and explicit dependencies on the standard library default to being
```
`cargo add` section says that it will automatically add `public = true` for built-in dependencies which was motivated by explicit dependencies defaulting to private.

epage commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1189 @@
+> any `builtin` dependencies it inferred. This allows the set of inferred
+> packages to change in the future if needed and prevents publishing a package
+> with a new Cargo from raising your MSRV. Similarly, the published `Cargo.toml`
+> will not explicitly declare any inferred dependencies.
```
Instead of tracking what is explicit or not, we can check if it matches the implicit state and strip it. We could then make them all explicit in the next edition.

ditto for `cargo metadata`

davidtwco replied:
I've changed this wording in 45f1a61f which I think addresses this concern

davidtwco committed:
fix mangled line wrapping

davidtwco committed:
clarify that patching is perma-unstable

davidtwco committed:
clarify public/private dependencies

davidtwco committed:
cargo doesn't force new lockfile versions

davidtwco committed:
add another reference to patching rationale

davidtwco committed:
clarify relevance to cargo index

davidtwco committed:
clarify patches are internal-use in unresolved q

davidtwco committed:
clarify cargo metadata back compat

davidtwco committed:
compare w/ implicit state vs tracking explicitness

epage commented:
@rfcbot fcp concern redundant-no-std

See https://github.com/rust-lang/rfcs/pull/3875#discussion_r2670232120

Nemo157 commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1189 @@
+crates, just as with any other dependency.
+
+In addition, uses of the `#![no_std]` attribute could be migrated to denying a
+lint which would prevent use of items from `std`.
```
I would also want `no_std` always added if you have

```toml
[dependencies]
core.builtin = true
std.builtin = true
std.optional = true
```

boozook replied:
> I would also want no_std always added if you have

Excuse me, isn't it break test-build especially if was defined like '#![cfg_attr(not(test), no_std)]`?

Nemo157 replied:
No, the test-harness always pulls `std` in as a dependency, unconditional `#![no_std]` already works fine with `#[test]`.

davidtwco committed:
pass `-Zcrate-attr=no_std`

Turbo87 commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1189 @@
+[rationale-cargo-index]: #why-add-standard-library-crates-to-cargos-index
+
+When Cargo builds the dependency graph, it is driven by the index (not
+`Cargo.toml`), so builtin dependencies need to be included in the index.
```
from the crates.io perspective, if the cargo team finds a way to make this work without index changes, then I think we'd prefer that. I don't think we have any major blockers with the proposed solution either, but if we can simplify it to not need index changes then even better.

since this discussion is apparently still ongoing I'll delay my FCP approval for now until I know what solution we ultimately will choose :)

epage replied:
We've already downloaded everything by the time we are building the unit graph and have access to all of the parsed manifests.  We already will have to deal with this for existing index entries / entries in the default state.

It might make things a little easier and cleaner.  It would also offer a simple programmatic way for end-users and linters to determine what is no-std and do no-std packages have std dependencies.

Should we leave this as an Unresolved Question for stabilization?

davidtwco replied:
Added an unresolved question in c0120b36

epage replied:
Thanks!

davidtwco committed:
build-dependencies is a two-way door

davidtwco committed:
elaborate on compatibility concern

davidtwco committed:
elaborate on cargo index need

programmerjake commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1271 @@
+>
+> The package version of `std`, `alloc` and `core` will be fixed at `0.0.0`. The
+> optional lockfile fields `dependencies` and `checksum` will not be present for
+> `builtin` dependencies.
```
they probably mean the version of rust that a particular `std` came with

davidtwco committed:
update rfcs#3874 links

davidtwco committed:
add unresolved question for cargo index changes

weihanglo commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1284 @@
+> [!NOTE]
+>
+> Explicit dependencies are passed to rustc without the `noprelude` modifier
+> ([?][rationale-explicit-noprelude]) (`noprelude` refers to the compiler's
```
The noprelude here and in the QA section are not clear to non-compiler experts. Wonder if it is worth linking to <https://doc.rust-lang.org/nightly/unstable-book/compiler-flags/extern-options.html#--extern-options> or elsewhere.

davidtwco replied:
Added a link in 3eb3aa61

weihanglo commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1284 @@
+[rfcs#3874-standard-library-crate-stability]: https://github.com/davidtwco/rfcs/blob/build-std-part-two-always/text/3874-build-std-always.md#standard-library-crate-stability
+[rfcs#3874-noprelude]: https://github.com/davidtwco/rfcs/blob/build-std-part-two-always/text/3874-build-std-always.md#why-use-noprelude-with---extern
+
+### Non-`builtin` standard library dependencies
```
Do we want to support [multiple locations](https://doc.rust-lang.org/nightly/cargo/reference/specifying-dependencies.html#multiple-locations)? E.g.,

```toml
[package]
name = "hello_world"
version = "0.1.0"
edition = "2024"

[dependencies]
std = { builtin = true, path = "../my_std" }
```

So the package remains publishable.

I don't know if there is really a use case for this though.

davidtwco replied:
I've added an explicit example of this in 2d2c6dbd. I intended for this to be supported as in this section we say that you can still support `path`/`git` for the standard library crates, so I don't see any reason why we wouldn't permit that at the same time as `builtin` so it is publishable. Same as registry sources and `path`/`git` sources being combined.

weihanglo commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1284 @@
+- [`cargo test`][cargo-test]
+- [`cargo uninstall`][cargo-uninstall]
+- [`cargo update`][cargo-update]
+- [`cargo vendor`][cargo-vendor]
```
Just want to confirm if a package has
 
```toml
[dependencies]
std = { git = "…" }
```

the std from a git repo will be vendored as a normal git repo. I believe the answer is yes.

davidtwco replied:
Yes, the vendoring behaviour doesn't change with explicit dependencies

weihanglo commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1284 @@
+
+- `cargo pkgid` and `cargo build --message-format` will unconditionally use the
+  new `pkgid` spec anyway;
+- `cargo_metadata` does not currently parse the pkgid spec so this won't break
```
Better to link to https://crates.io/crates/cargo_metadata to make explicit we are talking about the crate not the command.

davidtwco replied:
Added link in 3db44263

weihanglo commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1284 @@
+of standard library crates (see
+[build-std-always][rfcs#3874-standard-library-crate-stability]). Any explicit
+standard library dependency present in any dependency table will disable the
+implicit dependencies (e.g. an explicit `builtin` or `path` dependency from
```
This is a bit ambiguous. On line 170 the example shows that if a target has explicit std, implicit is disabled, while here "any" sounds like "any table including not my target platform dependencies table".


Maybe something like

```
Any explicit standard library dependency present in a dependency table
applicable to a given target will disable the implicit dependencies
for that target.
```

Or I understand wrong.

davidtwco replied:
Good catch, fixed in 456c2a9a

weihanglo commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1284 @@
+
+[patch.builtin] # permitted on nightly
+std = { .. }
+
```
Duplicate?

davidtwco replied:
Fixed in ee09558a

weihanglo commented on text/3875-build-std-explicit-dependencies.md:
Workspace inheritance of the `builtin` key isn't mentioned. I guess it could be left as an unresolved question, and may be fine just disallowing it in the initial version.

davidtwco replied:
I wasn't aware of this feature actually. I've added a brief mention in 90197173. I expect that after dependencies defined in the workspace are inherited, the same behaviour and constraints would apply as if they'd been written in the package directly.

weihanglo commented on text/3875-build-std-explicit-dependencies.md:
````diff
@@ -0,0 +1,1284 @@
+> ```
+
+[`cargo pkgid`][cargo-pkgid] when passed `-p core` would print
+`builtin://.#core` as the source, likewise with `alloc` and `std`. This format
````
The actual format can be deferred, right? This is best matching the `source` field in Cargo.lock to simplify implementation as well as a more consistent view of these formats.

davidtwco replied:
I've added an unresolved question in 927af0e5 noting that this format isn't fixed by the RFC and can be changed prior to stabilisation

weihanglo reviewed (commented):
Thanks and sorry for the late review.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3875/244eb0bb68bd5f3cc880cdd518d1345e28a8a627..c0120b3612567855f171e9fd79690e42b354757a)*

weihanglo commented:
@rfcbot reviewed

arlosi commented on text/3875-build-std-explicit-dependencies.md:
```diff
@@ -0,0 +1,1284 @@
+### Registries
+[registries]: #registries
+
+Standard library dependencies will be present in the registry index such that
```
Do we need to consider what happens when a crate with `builtin` deps is uploaded to an existing 3rd party registry without explicit support for it?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3875/244eb0bb68bd5f3cc880cdd518d1345e28a8a627..c0120b3612567855f171e9fd79690e42b354757a)*

davidtwco replied:
I'm not sure that this is any different than any other change or evolution of the registry index format such that it needs addressed specifically in this RFC. As I understand it, the format has been evolved in the past so could be again and there are means for doing so?

davidtwco committed:
link to rustc extern opts docs

davidtwco committed:
elaborate on combined sources

davidtwco committed:
add `cargo_metadata` crate link

davidtwco committed:
clarify applicability of dependency tables

davidtwco committed:
remove duplicate defn in example

davidtwco committed:
mention `workspace.dependencies`

davidtwco committed:
add unresolved q for pkgid format

davidtwco commented:
`@rust-rfcbot` is [having some issues](https://rust-lang.zulipchat.com/#narrow/channel/122651-general/topic/rfcbot.20down/near/590343372), so leaving this comment to note that all but two checkboxes in the FCP have been checked, which means this RFC should be entering its 10 day final comment period.

djc commented:
Sorry for the late feedback: given that Merriam-Webster lists "built-in" as the preferred spelling, should the source in `Cargo.toml` be spelled `built-in` instead of `builtin`?

bushrat011899 commented:
> Sorry for the late feedback: given that Merriam-Webster lists "built-in" as the preferred spelling, should the source in `Cargo.toml` be spelled `built-in` instead of `builtin`?

[`compiler-builtins`](https://github.com/rust-lang/compiler-builtins) implies the project has determined _builtin_ is close enough to a word, so I'd vote for that.

ehuss committed:
Set the tracking issue
